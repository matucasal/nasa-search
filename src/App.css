@import url("https://fonts.googleapis.com/css2?family=Montserrat&display=swap");
@import url("https://fonts.googleapis.com/css2?family=Play&display=swap");
* {
  margin: 0;
  padding: 0;
  -webkit-box-sizing: border-box;
          box-sizing: border-box;
}

body {
  font-family: 'Montserrat', sans-serif;
}

html {
  font-size: 62.5%;
  scroll-behavior: smooth;
}
/*# sourceMappingURL=App.css.map */