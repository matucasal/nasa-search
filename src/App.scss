//Import la typo
@import url('https://fonts.googleapis.com/css2?family=Montserrat&display=swap');
@import url('https://fonts.googleapis.com/css2?family=Play&display=swap');


* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

body{
    font-family: 'Montserrat', sans-serif;
}

html {
    font-size: 62.5%;
    scroll-behavior: smooth;
}