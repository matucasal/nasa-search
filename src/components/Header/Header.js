import './Header.scss';

import React, { Component } from 'react';
import Button from 'react-bootstrap/Button';
import { Person } from 'react-bootstrap-icons';

class Header extends Component {
    
    clickMyAccount() {
        console.log('Account clicked');
    }
    
    render() {
        return (
            <header>

                <div className="top-header">
                    
                    <div className="logo">
                        <img src='NASA_logo.png' alt='nasa-logo'></img>
                        <h1>NASA SEARCH</h1>
                    </div>
                    <div className="actions-header">
                        
                        <Button variant="outline-secondary" onClick={this.clickMyAccount}>
                            <Person className="account-icon"/> <span className="account-text">My Account</span>
                        </Button>
                    
                    </div>
                </div>
            </header>
        );
    }
   }
   
   export default Header;