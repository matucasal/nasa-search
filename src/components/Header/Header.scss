@import "../../styles/globalvars";


.top-header{
    @include flex(space-between, center, row);
    min-height: 10vh;
    flex-wrap: wrap;

    .logo{
        border: 1px solid red;
        font-family: 'Play', sans-serif;
        flex: 0 0 50%;
        @include flex(flex-start, center, row);
        h1{
            margin: 1rem;
            font-size: 10rem;
        }
        img{
            flex: 0 0;
            object-fit: cover;
        }
    
    }
    
    .actions-header{
        border: 1px solid blue;
        flex: 0 1 50%;
        @include flex(flex-end, center, row);
        button{
            border: 1px solid green;
            margin: 1rem;
        }
        
    
    }
}


@media screen and (max-width:950px){
    .top-header{
        .logo{
            flex: 0 0 70%;
            h1{
                font-size:3rem;
            }
        }
        .actions-header{
            flex: 0 0 30%;
        }
    }

    .account-text{
        display: none;
    }
}



