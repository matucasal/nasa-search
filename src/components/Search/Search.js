import './Search.scss';
import React, { Component } from 'react';
import { useNavigate, createSearchParams } from "react-router-dom";




const isEmpty = (str) => {
    return (!str || str.length === 0 );
}


class Search extends Component {
    
    

    state = {
        results: [],
        errors: {
            query_seach:'',
            year_start: '',
            year_end:''
        },
    }

    constructor(props) {
        super(props);

        this.state = {
            query_seach:'',
            year_start: '',
            year_end: '',
            errors: {
                query_seach:'',
                year_start: '',
                year_end:''
            },
        };


        this.onHandleQueryChange = this.onHandleQueryChange.bind(this);
        this.onHandleYearChange = this.onHandleYearChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);

        
    }

    /*Use this function to get lenght of number*/
    getLength(number) {
        return number.toString().length;
    }

    onHandleQueryChange= e => {
        let query = e.target.value;
        this.setState({
            [e.target.name]: query
        });
    };

    onHandleYearChange = e => {
        let year = e.target.value;

        //if not number, not allowed
        if (year !== '' && !Number(year)) {
            return;
        }

        if (year !== '' && this.getLength (year)>4 ){
            return;
        }

        //If year > this year -> error
        if (year !== '' && year > new Date().getFullYear()){
            this.state.errors[e.target.name] = "Year cannot be after " +  new Date().getFullYear();
        }
        else if (year !== '' && year < 1900){
            this.state.errors[e.target.name] = "Year cannot be before 1900";
        }
        else {
            this.state.errors[e.target.name] = '';
        }

        

        this.setState({
            [e.target.name]: year,
            errors: this.state.errors
        });
    };

    handleSubmit = event => {
        event.preventDefault();
        
        let validation = true

        //Validations
        //1) Query cannot be empty
        if (isEmpty(this.state.query_seach)){
            this.state.errors['query_seach'] = "The query cannot be empty";
            //console.log('this.state.errors["query_seach"]',this.state.errors["query_seach"])
            this.setState({ errors: this.state.errors });
            validation = false
        }
        //2) Validate the year end not < than year start
        if (this.state.year_end < this.state.year_start ){
            this.state.errors['year_start'] = "The start year cannot be bigger than the end year";
            //console.log('this.state.errors["query_seach"]',this.state.errors["query_seach"])
            this.setState({ errors: this.state.errors });
            validation = false
        }
        
        //Validations where ok
        if (validation){
            //Set errors to null
            this.state.errors['query_seach'] = '';
            this.state.errors['year_start'] = '';
            this.state.errors['year_end'] = '';

            this.setState({ errors: this.state.errors });
            let params =
            {
                q:this.state.query_seach, 
                media_type:'image',
                ...(this.state.year_start && {year_start: this.state.year_start}),
                ...(this.state.year_end && {year_end: this.state.year_end}),
            }

            console.log('params query input', params)
           
            this.props.navigate({
                pathname: '/results',
                search: `?${createSearchParams(params)}`,
            });
        }


    };
    
    render() {
        return (
            <div>
                <div className="search">
                    <form id="search-form" onSubmit={this.handleSubmit}>
                        <div>
                            <input className="input-search-query" 
                            type="text" 
                            name="query_seach" 
                            id="query_seach" 
                            placeholder="Search"
                            value = {
                                this.state.query_seach
                            }
                            onChange = {
                                this.onHandleQueryChange
                            } 
                            />
                            <span style={{ color: "red" }}>{this.state.errors["query_seach"]}</span>
                            
                        </div>
                        <div>
                            <input className="input-year-start" 
                            type="text" 
                            name="year_start" 
                            id="year_start" 
                            value = {
                                this.state.year_start
                            } 
                            placeholder="Year Start" 
                            onChange = {
                                this.onHandleYearChange
                            }/>
                            <span style={{ color: "red" }}>{this.state.errors["year_start"]}</span>
                        </div>

                        <div>
                            <input className="input-year-end" 
                            type="text" 
                            name="year_end" 
                            id="year_end"
                            value = {
                                this.state.year_end
                            } 
                            placeholder="Year End"
                            onChange = {
                                this.onHandleYearChange
                            }
                            />
                            <span style={{ color: "red" }}>{this.state.errors["year_end"]}</span>
                        </div>
                    </form>
                    <button type="submit" form="search-form" value="Submit">Seach</button>
                </div>

               
                
            </div>
        );
    }
   }



//Export this way to use navigate
export default (props) => (
    <Search
        {...props}
        navigate={useNavigate()}
    />
);




