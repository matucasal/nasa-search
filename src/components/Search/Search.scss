@import "../../styles/globalvars";

.search{
    @include flex(space-between, center, column);
    min-height: 10vh;
    margin-top: 5vh;
    width: 95%;
    flex-wrap: wrap;
    
    form{
        display: flex;
        flex-direction: row;
        margin-left: 1rem;
        .input-search-query{
            box-sizing: border-box;
            cursor: text;
            display: flex;
            background-color: rgb(255, 255, 255);
            border-radius: 10px;
            border: 1px solid rgb(229, 232, 235);
            width: 100%;
            min-width: 80rem;
            padding: 1rem;
            box-shadow: rgba(4, 17, 29, 0.25) 0px 0px 8px 0px;
            height: 4rem;
        }

        
        

        .input-year-start
        {
            box-sizing: border-box;
            cursor: text;
            display: flex;
            background-color: rgb(255, 255, 255);
            border-radius: 10px;
            border: 1px solid rgb(229, 232, 235);
            width: 100%;
            min-width: 10rem;
            padding: 1rem;
            box-shadow: rgba(4, 17, 29, 0.25) 0px 0px 8px 0px;
            height: 4rem;
            
        }
        .input-year-end
        {
            box-sizing: border-box;
            cursor: text;
            display: flex;
            background-color: rgb(255, 255, 255);
            border-radius: 10px;
            border: 1px solid rgb(229, 232, 235);
            width: 100%;
            min-width: 8rem;
            padding: 1rem;
            box-shadow: rgba(4, 17, 29, 0.25) 0px 0px 8px 0px;
            height: 4rem;
        }
    }

    button{
        margin-top:2rem;
        box-sizing: border-box;
        flex-direction: row;
        color: rgb(255, 255, 255);
        text-decoration: none;
        display: inline-flex;
        -webkit-box-align: center;
        align-items: center;
        border-radius: 10px;
        -webkit-box-pack: center;
        justify-content: center;
        font-size: 16px;
        font-weight: 600;
        padding: 12px 20px;
        background-color: rgb(24, 104, 183);
        border: 1px solid rgb(32, 129, 226);
        transition: all 0.2s ease 0s;
        box-shadow: rgba(4, 17, 29, 0.25) 0px 0px 8px 0px;
	    width: 20rem;
    }
  
}


@media screen and (max-width:950px) and (min-width:300px){
    .search{
        form{
            .input-search-query{
                min-width: 25rem;
            }
            .input-year-start{
                min-width: 5rem;
            }
            .input-year-end{
                min-width: 5rem;
            }
        }
    }
}


@media screen and (max-width:300px){
    .search{
        form{
            .input-search-query{
                min-width: 25rem;
            }
            .input-year-start{
                display:none;
            }
            .input-year-end{
                display:none;
            }
        }
    }
}

