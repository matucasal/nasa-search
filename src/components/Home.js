import React, { Component } from 'react';
import Header from './Header/Header';
import Search from './Search/Search';
import './Home.css'

class Home extends Component {
    render() {
      return (
        <div>
            <Search></Search>
            <hr className="rounded"></hr>
        </div>
      );
    }
}
   
export default Home;