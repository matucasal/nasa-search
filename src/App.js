import './App.scss';
import  { HashRouter, Routes, Route } from "react-router-dom";

import ShowPage from './components/ShowPage/ShowPage';
import Home from './components/Home'

function App() {
  return (
    <div className="App">
        <HashRouter>
            <div>
                
                <div className="content">
                    <Routes>
                    <Route path='/' element={<Home/>}/>
                        <Route path='/results' element={<ShowPage/>}/>
                        <Route path='/show' element={<ShowPage/>}/>
                    </Routes>
                </div>
            </div>
        </HashRouter>
    </div>
  );
}

export default App;
