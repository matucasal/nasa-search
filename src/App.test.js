// pay attention to write it at the top level of your file
//Using this to create the router "context"
const mockedNavigate = jest.fn();

import {render, fireEvent, waitFor, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from './App';
import Header from './components/Header/Header'
import Home from './components/Home'
import Search from './components/Search/Search'



/*
What to test ->
    0) Must render
    1) Render the right screen
    2) Inputs -> Form Validation
    3) test the events of the form
    4) Search input try to search
    5) Test the props recieved by the search in the results page
    6) Test the states of the results page
    7) Test loading in the results page
    8) Select an item try to enter the show page
    9) Array manipulation for images in the show page
*/


Enzyme.configure({adapter: new Adapter()});

//0) Must render
it('renders correctly the app', () => {
    const tree = renderer
    .create(<App></App>)
    .toJSON();
    expect(tree).toMatchSnapshot();
});


//1) Render the right screen -> Header / Home 
describe('Parent Component App', () => {
    it('renders Child component Header and Home', () => {
        const wrapper = shallow(<App/>);
        expect(wrapper.containsMatchingElement(<Header />)).toEqual(true);
        expect(wrapper.containsMatchingElement(<Home />)).toEqual(true);
    });
});


//Using this to create the router "context"
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedNavigate
}));



//2) Inputs -> Form Validation
describe("Testing Search", () => {
 
    test('search input exists and correct type', () => {
      render(<Search />);
      const inputEl = screen.getByTestId("search-input");
      expect(inputEl).toBeInTheDocument();
      expect(inputEl).toHaveAttribute("type", "text");
    });

    test('year strart input exists and correct type', () => {
        render(<Search />);
        const inputEl = screen.getByTestId("year-start-input");
        expect(inputEl).toBeInTheDocument();
        expect(inputEl).toHaveAttribute("type", "text");
    });

    test('year end input exists and correct type', () => {
        render(<Search />);
        const inputEl = screen.getByTestId("year-end-input");
        expect(inputEl).toBeInTheDocument();
        expect(inputEl).toHaveAttribute("type", "text");
    });

    test('year start input to be ok', () => {
        render(<App />);
     
        const inputElYearStart = screen.getByTestId("year-start-input");
        userEvent.type(inputElYearStart, "2020");

        const inputElYearEnd = screen.getByTestId("year-end-input");
        userEvent.type(inputElYearEnd, "2019");
     
        expect(screen.getByTestId("year-start-input")).toHaveValue("2020");
        expect(screen.queryByTestId("year-start-error-msg")).not.toBeInTheDocument();
    });

})
