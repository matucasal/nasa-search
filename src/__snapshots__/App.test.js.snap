// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly the app 1`] = `
<div
  className="App"
>
  <div>
    <header>
      <div
        className="top-header"
      >
        <div
          className="logo"
        >
          <img
            alt="nasa-logo"
            src="NASA_logo.png"
          />
          <h1>
            NASA SEARCH
          </h1>
        </div>
        <div
          className="actions-header"
        >
          <button
            className="btn btn-outline-secondary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            <svg
              className="account-icon"
              fill="currentColor"
              height="1em"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M8 8a3 3 0 1 0 0-6 3 3 0 0 0 0 6zm2-3a2 2 0 1 1-4 0 2 2 0 0 1 4 0zm4 8c0 1-1 1-1 1H3s-1 0-1-1 1-4 6-4 6 3 6 4zm-1-.004c-.001-.246-.154-.986-.832-1.664C11.516 10.68 10.289 10 8 10c-2.29 0-3.516.68-4.168 1.332-.678.678-.83 1.418-.832 1.664h10z"
              />
            </svg>
             
            <span
              className="account-text"
            >
              My Account
            </span>
          </button>
        </div>
      </div>
    </header>
    <div>
      <div>
        <div
          className="search"
        >
          <form
            id="search-form"
            onSubmit={[Function]}
          >
            <div>
              <input
                className="input-search-query"
                data-testid="search-input"
                id="query_seach"
                name="query_seach"
                onChange={[Function]}
                placeholder="Search"
                type="text"
                value=""
              />
              <span
                data-testid="query-search-error-msg"
                style={
                  Object {
                    "color": "red",
                  }
                }
              />
            </div>
            <div>
              <input
                className="input-year-start"
                data-testid="year-start-input"
                id="year_start"
                name="year_start"
                onChange={[Function]}
                placeholder="Year Start"
                type="text"
                value=""
              />
            </div>
            <div>
              <input
                className="input-year-end"
                data-testid="year-end-input"
                id="year_end"
                name="year_end"
                onChange={[Function]}
                placeholder="Year End"
                type="text"
                value=""
              />
              <span
                data-testid="year-end-error-msg"
                style={
                  Object {
                    "color": "red",
                  }
                }
              />
            </div>
          </form>
          <button
            form="search-form"
            type="submit"
            value="Submit"
          >
            Seach
          </button>
        </div>
      </div>
      <hr
        className="rounded"
      />
    </div>
    <div
      className="content"
    >
      <div
        className="search-results"
      >
        There are no results to show.
      </div>
    </div>
  </div>
</div>
`;
